test/e2e/apimachinery/watch.go

```diff
@@ -0,0 +1,209 @@
+			p.Spec.ActiveDeadlineSeconds = int64ptr(5)
+		})
+		Expect(err).NotTo(HaveOccurred())
+		expectEvent(watchA, watch.Modified, testPodA)
+		expectEvent(watchAB, watch.Modified, testPodA)
+		expectNoEvent(watchB, watch.Modified, testPodA)
+
+		err = c.CoreV1().Pods(ns).Delete(testPodA.GetName(), nil)
+		Expect(err).NotTo(HaveOccurred())
+		expectEvent(watchA, watch.Deleted, nil)
+		expectEvent(watchAB, watch.Deleted, nil)
+		expectNoEvent(watchB, watch.Deleted, nil)
+
+		testPodB, err = c.CoreV1().Pods(ns).Create(testPodB)
+		Expect(err).NotTo(HaveOccurred())
+		expectEvent(watchB, watch.Added, testPodB)
+		expectEvent(watchAB, watch.Added, testPodB)
+		expectNoEvent(watchA, watch.Added, testPodB)
+
+		err = c.CoreV1().Pods(ns).Delete(testPodB.GetName(), nil)
+		Expect(err).NotTo(HaveOccurred())
+		expectEvent(watchB, watch.Deleted, nil)
+		expectEvent(watchAB, watch.Deleted, nil)
+		expectNoEvent(watchA, watch.Deleted, nil)
+	})
+})
+
+func watchPodsWithLabels(f *framework.Framework, labels ...string) (watch.Interface, error) {
+	c := f.ClientSet
+	ns := f.Namespace.Name
+	opts := metav1.ListOptions{
+		LabelSelector: metav1.FormatLabelSelector(&metav1.LabelSelector{
+			MatchExpressions: []metav1.LabelSelectorRequirement{
+				{
+					Key:      watchPodLabelKey,
+					Operator: metav1.LabelSelectorOpIn,
+					Values:   labels,
+				},
+			},
+		}),
+	}
+	return c.CoreV1().Pods(ns).Watch(opts)
+}
+
+func int64ptr(i int) *int64 {
+	i64 := int64(i)
+	return &i64
+}
+
+type updatePodFunc func(p *v1.Pod)
+
+func updatePod(f *framework.Framework, name string, update updatePodFunc) (*v1.Pod, error) {
+	c := f.ClientSet
+	ns := f.Namespace.Name
+	var p *v1.Pod
+	pollErr := wait.PollImmediate(2*time.Second, 1*time.Minute, func() (bool, error) {
+		var err error
+		if p, err = c.CoreV1().Pods(ns).Get(name, metav1.GetOptions{}); err != nil {
+			return false, err
+		}
+		update(p)
+		if p, err = c.CoreV1().Pods(ns).Update(p); err == nil {
+			return true, nil
+		}
+		// Only retry update on conflict
+		if !errors.IsConflict(err) {
+			return false, err
+		}
+		return false, nil
+	})
+	return p, pollErr
+}
+
+func expectEvent(w watch.Interface, eventType watch.EventType, object runtime.Object) {
+	if event, ok := waitForEvent(w, eventType, object); !ok {
+		framework.Failf("Timed out waiting for expected event: %v", event)
+	}
+}
+
+func expectNoEvent(w watch.Interface, eventType watch.EventType, object runtime.Object) {
+	if event, ok := waitForEvent(w, eventType, object); ok {
+		framework.Failf("Unexpected event occurred: %v", event)
+	}
+}
+
+func waitForEvent(w watch.Interface, eventType watch.EventType, object runtime.Object) (watch.Event, bool) {
+	stopTimer := time.NewTimer(10 * time.Second)
+	expect := watch.Event{
+		Type:   eventType,
+		Object: object,
+	}
+	for {
+		select {
+		case actual := <-w.ResultChan():
+			if expect.Type == actual.Type && (expect.Object == nil || apiequality.Semantic.DeepEqual(expect.Object, actual.Object)) {
```

Comment:
When running this e2e test locally it fails with the error: `"Timed out waiting for expected event: {DELETED <nil>}"`. The comparison `expect.Object == nil` doesn't evaluate to true in the case of delete as we would expect, and this leads to a timeout. This is because `expect.Object` is an interface whose underlying value is nil but the interface itself is non-nil (https://golang.org/doc/faq#nil_error). Replacing `expect.Object` with `object` will fix this :)